src/nvim/path.c

```diff
@@ -612,8 +612,10 @@ static size_t do_path_expand(garray_T *gap, const char_u *path,
   *s = NUL;
 
   Directory dir;
-  if (os_scandir(&dir, *buf == NUL ? "." : (char *)buf)
-      || os_isdir(*buf == NUL ? (char_u *)"." : (char_u *)buf)) {
+  char *dirpath = (*buf == NUL ? "." : (char *)buf);
+  if (os_file_is_readable(dirpath)
+      && (os_scandir(&dir, dirpath)
+          || os_isdir((char_u *)dirpath))) {
```

Comment:
The segfault, I believed, was caused by `os_scandir()` returning `false`, freeing its internal memory, `os_isdir()` returning `true`, then `scandir_next_with_dots()` reading the invalid `Directory` object. In my testing, uv's `scandir` doesn't know the difference between an empty directory and one it doesn't have permissions to read, so making `os_scandir()` return `true` when `uv_fs_scandir()` returns `0` and not free its memory should fix the segfault. (`uv_fs_scandir()` returns zero when the directory is empty, and when it doesn't have permissions--in either case, `os_scandir_next()` will be safe to call.)

The `os_isdir()` call is a serious problem. Please modify `os_scandir()` to fix this bug and remove the isdir check. The `os_file_is_readable()` check is good, though, so we don't return `"."` and `".."` on non-readable directories (which is consistent with vanilla vim).